Prune more aggressively during IBD

Sjors opened the pull request:
Pruning forces a chainstate flush, which can defeat the dbcache and harm performance significantly.

During IBD we now prune based on the worst case size of the remaining blocks, but no further than
the minimum prune size of 550 MB.

Using `MAX_BLOCK_SERIALIZED_SIZE` is complete overkill on testnet and usually too high on mainnet. It doesn't take into account the SegWit activation block either. This causes the node to be further pruned than strictly needed after IBD. It also makes it more difficult to test. One improvement could be to use a moving average actual block size or a hard coded educated guess. However there's something to be said for keeping this simple.

Sjors commented:
@fanquake probably also needs "Block storage" label.

esotericnonsense commented:
Untested ACK, would kill off #11658 and #11359.

luke-jr commented on src/validation.cpp:
```diff
@@ -3570,6 +3570,9 @@ static void FindFilesToPrune(std::set<int>& setFilesToPrune, uint64_t nPruneAfte
 
     unsigned int nLastBlockWeCanPrune = chainActive.Tip()->nHeight - MIN_BLOCKS_TO_KEEP;
     uint64_t nCurrentUsage = CalculateCurrentUsage();
+    // Worst case remaining block space:
```
Seems like this ought to be using best case...?

Sjors replied:
Best case would be empty blocks. That would lead to a lot of flushes.

morcos commented:
ACK 86bef23

eklitzke commented:
utACK 86bef23e6550cdcf989ae6ac22dbbc45bbf613e4

Sjors commented:
Rebased due to release notes change.

Sjors commented:
Rebased due to release notes change.

Sjors commented:
`p2p_leak.py` seems a bit random (and passes on my local machine)...

eklitzke commented:
utACK 82efbf1e8ac67ad9d04cba9b64cb79ece86209f8

luke-jr commented:
Before merging, please remove the name and PR reference from the commit message, so it doesn't ping us every time someone adds it to their random fork.

neverstopthegrind1 commented:
I found this web in my cache log why ?

On Sat, Mar 31, 2018, 4:27 PM Luke Dashjr <notifications@github.com> wrote:

> Before merging, please remove the name and PR reference from the commit
> message, so it doesn't ping us every time someone adds it to their random
> fork.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/12404#issuecomment-377721048>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AkND8slkw6vU8yLJBAdXMe9z10A-G6ddks5tj-bLgaJpZM4SA7zL>
> .
>

Sjors commented:
@luke-jr will do. Should I also remove it from the PR description, since that also ends up in the merge commit message? Or do those merge commits rarely make it into upstream work because commits are cherry-picked?

Sjors committed:
Prune more aggressively during IBD

Pruning forces a chainstate flush, which can defeat the dbcache and harm performance significantly.

During IBD we now prune based on the worst case size of the remaining blocks, but no further than
the minimum prune size of 550 MB.

Sjors commented:
Done. Also: rebased for release notes.

Sjors commented:
I've been racing AWS instances for the past few days, using `master`, #11658 (rebased on master) and this PR. I use a `t2.micro` with 1 vCPU, 1 GiB RAM and 20 GB storage. I set `prune` to 10 GB, `dbcache=300` and `maxmempool=5`.

After 72 hours master is currently at block 341909, @luke-jr's branch is at 364905 and mine is at 360719. 

I enabled T2 Unlimited to prevent CPU throttling, although it doesn't seem to be CPU bound beyond the first 100-200K blocks (that will change after the assumed valid block):
<img width="882" alt="schermafbeelding 2018-05-20 om 10 42 26" src="https://user-images.githubusercontent.com/10217/40277140-920cbbf4-5c1a-11e8-9116-eb84bc45e4a8.png">

I tried higher values for `dbcache` but that led to out of memory crashes (sometimes during a cache flush) and once even to a machine freezing.

I'll leave them running for a bit. So far it seems clear that merging either of these PR's would be quite helpful on low-end machines, but which one is less clear. It probably depends on the choices for `dbcache` and `prune` and my guess is that machines with more RAM would benefit from pruning as aggressively as possible to minimize the number of cache flushes.

Sjors commented:
To clarify, is `IsInitialBlockDownload()` something that only happens once in the life time of a node, or is this also true if it needs to do a large catch up? If the latter, there is a case to be made for conservative pruning (or putting aggressive pruning behind a config flag).

When you run something like c-lightning against a pruned bitcoind node, it's constantly processing blocks as they come in. A large prune event could mess with this process if for some reason the other application isn't completely caught up. This is less if a problem for the initial sync if that other application doesn't care about pre-history. E.g. c-lightning it doesn't need history before its wallet creation block, so the trick there is to wait with launching it until `bitcoind` finishes IBD, and then keep them both running.

But there may be other applications that need to track some sort of index all the way through IBD where it's important they don't lose sync.

Sjors commented:
After a little under 24 hours the `t2.medium` instances:

* master: 458269
* 10% pruning: 471894
* this PR: 396051

Notice how this PR so far seems to perform _worse_ than master. I'll keep an eye on it. Maybe it has something to do with the large `dbcache`? Because of the more frequent prune events the `dbcache` doesn't grow as much on master and the 10% prune branch. See [IRC](https://botbot.me/freenode/bitcoin-core-dev/2018-05-21/?msg=100272576&page=1). Paging @eklitzke.

Sjors commented:
```sh
echo "height, cache" > cache.csv
cat ~/.bitcoin/debug.log | grep UpdateTip | awk -F'[ =M]'  '{print $7", " $19 }' >> cache.csv
```

<img width="1148" alt="2018-05-21 10 56 57" src="https://user-images.githubusercontent.com/10217/40299195-cb706a6e-5ce5-11e8-815e-6d749fb81132.png">

Sjors commented:
This extracts block height, cache size and a unix timestamp from the log: 
```sh
cat prune300_master.log | grep UpdateTip | gawk -F'[ =M]'  '{print $7", " $19", " gsub(/[-T:Z]/," ") ", " mktime($1 " " $2 " " $3 " " $4 " " $5 " " $6)  }' >> prune300_master.csv
```

IBD duration with dbcache=300MB: 

<img width="806" alt="2018-05-25 17 39 59" src="https://user-images.githubusercontent.com/10217/40553385-c32fec52-6042-11e8-8412-280056fcaa19.png">

Vertical access is in days. They ran for more than a week but didn't finish. The 10% prune strategy (green line) was the fastest, master the slowest.

Cache usage:

<img width="769" alt="2018-05-25 17 40 18" src="https://user-images.githubusercontent.com/10217/40553386-c34a7e28-6042-11e8-9829-9b6c0e328229.png">

Both strategies use more cache than master, but don't differ much for such a small dbcache.

IBD duration with dbcache=300MB: 
<img width="820" alt="2018-05-25 17 40 25" src="https://user-images.githubusercontent.com/10217/40553388-c366d172-6042-11e8-8575-c25e4986ce50.png">

The 10% pruning strategy (green) was the only one that finished IBD before I gave up. This PR (red line) is dramatically slower than even master.

Cache usage:
<img width="793" alt="2018-05-25 17 40 31" src="https://user-images.githubusercontent.com/10217/40553389-c38792c2-6042-11e8-814e-07d1dee107be.png">

Both strategies use more cache than master. Aggressive pruning uses way more cache, but for some reason that seems to make things worse.

n1bor commented:
FYI on AWS been running a node with 4gig RAM and sc1 disks (very cheap - 0.025/GigMonth) for a node with txindex on and keeps up fine (i.e. does not use burst allowance). Is used by a lightning node so get reasonable number of rpc requests. Is useless for IBD, but can set to SSD initially then once IBD done switch to sc1 with the click of a button!

Sjors commented:
@n1bor for [my own project on AWS](https://medium.com/provoost-on-crypto/bitcoin-core-lightning-rails-on-aws-ad3bd45b11e0), I also use the strategy of doing IBD on a fast machine (`i3.2xlarge`). Anything with > 10 GB RAM to prevent the cache from flushing and a disk big enough to avoid pruning (doesn't have to be SSD). The bigger disk is ephemeral and goes away after you downgrade to a regular instance type, so I do a manual prune at the end and copy the data to the persistent disk.

I'll look into [Cold HDD (sc1) Volumes](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSVolumeTypes.html#EBSVolumeTypes_sc1) for that project, because it `c-lightning` doesn't fully love pruned nodes yet, but that's off topic...

But you don't have that luxury on a physical machine like the various [Rasberry / Orange / Nano] Pi's out there. So it's quite useful if IBD performed better on a constrained machines. Also, even that price point pruning still saves $5 a month in storage at the current chain size.

Sjors commented:
Zooming in a little bit, this branch started dramatically slowing down compared to master around block 375000, which is around the September 2015 UTXO set explosion:

<img width="138" alt="utxo 2015" src="https://user-images.githubusercontent.com/10217/41196016-39958e14-6c38-11e8-97f5-6402a15b47ad.png">

Perhaps the performance of read or write operations involving `CCoinsCacheEntry` with `DIRTY` flag dramatically decrease when cache is > ~ 1GB? That would explain why higher frequency pruning, which generally keeps cache below 300 MB in that period doesn't slow down. While at the same time it would explain why 10 GB of cache, where all entries are `FRESH`, doesn't slow down either. Does that seem even plausible?

I could test this by deliberately interrupting IBD on master on a non pruned node at roughly the same blocks where this branch flushed the cache: 244000, 298000, 332000, 355000, 373000, 388000, 401000, 414000, 424000, 436000, 446000, 455000 (the latter two being the range where master starts slowing down).

Sjors commented:
Related IRC discussion: https://botbot.me/freenode/bitcoin-core-dev/2018-06-09/?msg=100959313&page=2

<img width="1253" alt="schermafbeelding 2018-06-10 om 11 02 34" src="https://user-images.githubusercontent.com/10217/41200055-d536e286-6c9d-11e8-8b4b-239144dbb6f8.png">

n1bor commented:
@Sjors not sure if you ever saw this: https://gist.github.com/n1bor/d5b0330a9addb0bf5e0f869518883522
Feels to me that time spent on IBD for pruned nodes would be better spent on chainstate only download type solution. Factor of 50x speed up. But needs a softfork - so maybe not!

sipa commented:
@n1bor That seems orthogonal. Synchronizing from chainstate is a very interesting idea, but it's also a completely different security model (trusting that the historical chain has correct commitments rather than computing things yourself).

n1bor commented:
My take is we have on order of "goodness":
1) Full Node
2) Pruned Full Node
3) Chain-State Downloaded Full Node with soft-fork to commit chainstate to headers. (what my post was about)
4) SPV
5) Web-Wallets
Currently core on offers 1 & 2.

Just think if core offered 3 would reduce number of users using web-wallets/SPV. Which has only got to be a good thing.

sipa commented:
I agree that would be a good thing, but it in no way changes the fact that we should have a performant implementation for those who do not want to rely on trusting such commitments.

Also, this is not the place to discuss changes to the Bitcoin protocol.

Sjors commented:
I launched two new t2.medium nodes on AWS, running Ubuntu 16.04, 2 CPU (uncapped), 4 GB RAM no swap. I set `prune=10000`, `dbcache=3000` and `maxmempool=5` on both like I did earlier. The blue lines are the current master master, the orange line is this PR rebased on master.

Again, this branch slows down dramatically quite early on, this time I captured some metrics:

<img width="880" alt="cpu" src="https://user-images.githubusercontent.com/10217/41251712-1288111e-6dbb-11e8-8b10-74cdb2879f46.png">

<img width="879" alt="network" src="https://user-images.githubusercontent.com/10217/41251723-1784ce0a-6dbb-11e8-9dea-ec96248b8e92.png">

<img width="1732" alt="blue_is_master" src="https://user-images.githubusercontent.com/10217/41251738-24f62a8e-6dbb-11e8-9310-f466a80e3ac0.png">

There are prune events at 15:28 (block 244388, cache 929 MB), 15:38 (297332, cache 1024 MB), 14:48 (331446, cache 1235 MB), 15:58 (354941, cache 1279 MB) and 16:11 (373100, cache 2951 MB). Those last two are right before and after the network activity drops.

I'll try spinning up a node with `dbcache=1000`

I ran the same configuration on my iMac (which has 4 CPU's and a USB 3.1gen2 external SSD drive) and don't get any noticeable performance difference between these two branches (2 hours 20 minutes to run from block 360.000 to 480.000).

Sjors commented:
I don't see `LogPrint(BCLog::PRUNE, "Prune: target=%dMiB actual=%dMiB diff=%dMiB max_prune_height=%d removed %d blk/rev pairs\n"` appear in the logs, not even for master. That category isn't disabled by default, is it?

Sjors commented:
Running this branch with `dbcache=1000` doesn't cause the same high read disk activity:
<img width="891" alt="db cache 1000" src="https://user-images.githubusercontent.com/10217/41281426-a4498fd0-6e31-11e8-995f-945e9027b227.png">

It's still running so I don't know if it's faster than master or the 10% prune strategy, but at least it doesn't suffer a similar slow down as `dbcache=3000`.

Sjors commented:
The thick line shows this PR with dbcache set to 1000. It no longer shows the performance hit you see with dbcache=3000 and it's faster than master, but not necessarily faster than the 10% pruning strategy.

<img width="1119" alt="2018-06-14 17 10 14" src="https://user-images.githubusercontent.com/10217/41420915-d57b0aaa-6ff5-11e8-900b-44b28cfd3587.png">

Closing this in favor of #11658, since the benefit seems small and an unexplained massive performance hit needs... explaining :-)

ajtowns commented:
FWIW, one effect I'm seeing that might cause the difference between dbcache 3000 vs 1000 is that when the cache is flushed, it takes a little while (and presumably 3x as long with 3x as large a dbcache), during which the block download queues pretty much empty, and then after the cache is flushed, the queues take a while to even out and get back up to the same download speed.
